[Merged by Bors] - feat(topology): one-point compactification of a topological space

justadzr opened the pull request:
Define `alexandroff X` to be the one-point compactification of a topological space `X` and prove some basic lemmas about this definition.

Co-authored by: Yury G. Kudryashov <urkud@urkud.name>

---

Adding the one-point compactification of a topological space. This file firstly defines the type synonyms and adds all other instances. A discussion on whether the topology should be constructed on `option` or the type synonym is available [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/formalizing.20one.20point.20compactification). [This draft](https://github.com/justadzr/Lean_2021/blob/complex-diff/src/alexandroff'.lean) is a version that defines the topology on `option` and then makes it as an instance on the one-point extension.

The following things might be worth considering:
* There will be more trivial lemmas if we want to define **all** structures on `option` first and then put them on the one-point extension, as most of them are set-related and would still be useful after defining the type synonyms.
* If we only define the topology on `option` and make all other instances on the one-point extension, then the number of trivial lemmas would be the same as it is in this PR.

- [x] depends on: #8636
- [x] depends on: #8650 
- [x] depends on: #8651


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

justadzr committed:
one-point compactification of a topological space

justadzr committed:
change a name

justadzr committed:
change a name

justadzr committed:
remove one extra coersion

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,319 @@
+def of : X → alexandroff X := some
+
+/-- The range of the embedding -/
+def range_of (X : Type*) [topological_space X] : set (alexandroff X) := of '' (univ : set X)
```
Is there an advantage to having this vs just using `set.range of`?

justadzr replied:
Nope. Changed the definition to `range (@of X _)`. Is it possible to golf the new proof of `connected_space_alexandroff`? I have to use a `rw` in that proof because there is no lemma `is_preconnected_range` in mathlib.

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,319 @@
+
+instance : topological_space (alexandroff X) :=
+{ is_open := λ s, if ∞ ∈ s then is_compact (of⁻¹' s)ᶜ ∧ is_open (of⁻¹' s)
+    else is_open (of⁻¹' s),
```
Possibly easier to work with as
```suggestion
{ is_open := λ s, (∞ ∈ s → is_compact (of ⁻¹' s)ᶜ) ∧ is_open (of ⁻¹' s),
```
obviously the proofs below would have to change.

justadzr replied:
Would this force us to use more `by_cases` in the proof? With the former definition, we can use `split_ifs` quite conveniently.

eric-wieser replied:
I expect it would make things simpler because you no longer have to prove `is_open (of⁻¹' s)` twice

justadzr replied:
I still think the `if then else` definition is better. In most cases, `is_open (of⁻¹' s)` can be proved with two super short `exact` statements because the key arguments of the proofs can be done on `if... then... else...`, and we no longer need to write `by_cases h : ∞ ∈ s` and specialize the definition to get `is_compact (of ⁻¹' s)ᶜ`.

And for other proofs I use the simplified lemmas `is_open_iff_mem/not_mem`, which would be the same for both definitions.

eric-wieser replied:
Have a look at my suggested proofs - they all _remove_ `by_cases h : ∞ ∈ s` rather than adding it.

justadzr replied:
Wow they are super elegant.

justadzr committed:
change the defition of `range_of`

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,322 @@
+    { simpa [h, h'] using hs.inter ht.2 },
+    { simpa [h, h'] using hs.2.inter ht },
+    { simpa [h, h'] using hs.inter ht }
+  end,
```
As an example: with my suggestion above, this becomes
```suggestion
  is_open_inter := λ s t ⟨hsc, hs⟩ ⟨htc, ht⟩, begin
    refine ⟨_, hs.inter ht⟩,
    rintro ⟨hsi, hti⟩,
    simpa [compl_inter] using (hsc hsi).union (htc hti),
  end,
```

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,322 @@
+     split_ifs at ht,
+     { exact ht.right },
+     { exact ht }
+  end }
```
Avoiding `if` also makes this one shorter:
```suggestion
  is_open_sUnion := λ S ht, begin
    suffices : is_open (of⁻¹' ⋃₀S),
    { refine ⟨_, this⟩,
      intro h,
      obtain ⟨(a : set (alexandroff X)), ha, ha'⟩ := mem_sUnion.mp h,
      replace ht := (ht a ha).1 ha',
      refine compact_of_is_closed_subset ht this.is_closed_compl _,
      rw [compl_subset_compl, preimage_subset_iff],
      intros y hy,
      refine ⟨a, ha, hy⟩ },
    rw is_open_iff_forall_mem_open,
    simp only [and_imp, exists_prop, mem_Union, preimage_sUnion, mem_preimage, of_eq_coe,
               exists_imp_distrib],
    intros y s hs hy,
    refine ⟨of ⁻¹' s, subset_subset_Union _ (subset_subset_Union hs (subset.refl _)), _,
       mem_preimage.mpr hy⟩,
    exact (ht s hs).2,
  end }
```

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,322 @@
+  by_cases H : ∞ ∈ s,
+  { simpa using ((is_open_iff_of_mem H).mp h).2 },
+  { exact (is_open_iff_of_not_mem H).mp h }
+end
```
With my version this becomes
```suggestion
h.2
```

eric-wieser commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,322 @@
+  { rw is_open_iff_of_mem H at hs,
+    have minor₁ : s ≠ {∞},
+    { by_contra w,
+      rw [not_not.mp w, of_preimage_infty, compl_empty] at hs,
```
`by_contra` followed by `rw not_not` is just `intro`:
```suggestion
    { intro hw,
      rw [hw, of_preimage_infty, compl_empty] at hs,
```

eric-wieser replied:
Or even
```suggestion
    { rintro rfl,
      rw [of_preimage_infty, compl_empty] at hs,
```

justadzr replied:
Thank you for your reviews! I could learn a lot from them. What does `rintro rfl` do? I guess it's like saying intro some equality and register it as a `rfl`?

eric-wieser replied:
`rintro rfl` is like doing a pattern match on `| rfl :=` in the equation compiler. You can read it as roughly `intro w`, `subst w`.

justadzr replied:
Thank you!

justadzr committed:
changed the main definition

urkud commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,308 @@
+
+/-- The Alexandroff extension of an arbitrary topological space `X` -/
+@[nolint unused_arguments]
+def alexandroff (X : Type*) [topological_space X] := option X
```
Why do you need `[topological_space X]` in the definition of `alexandroff`?

justadzr replied:
Thanks for the review. Due to my very limited mathematical background, I have the impression that people would never use `alexandroff` if `X` is not a topological space. But I am not sure if I should define `alexandroff` with this argument in Lean, as things like `stone_cech X` rely on `topological_space X` but the definition of `with_top` includes no such arguments.

urkud replied:
Yes, my question was about Lean definition, not math definition. As you can see, the definition of `stone_cech X` uses the argument `[topological_space X]`. This is the difference between `stone_cech` and `with_top`. In cases like `with_top`/`with_bot`/`with_one`/`alexandroff` we avoid unused arguments.

There is one important case when we allow unused arguments in type synonyms, that is when some structures on the new type use these arguments. E.g., `category_theory.single_obj (M : Type*) := unit` is a one-object category with `Hom(*, *) = M` and `f ≫ g = g * f`.

justadzr replied:
Thank you for the detailed explanation! I will modify the code.

urkud commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,308 @@
+
+instance : has_coe_t X (alexandroff X) := ⟨of⟩
+
+instance : inhabited(alexandroff X) := ⟨∞⟩
```
```suggestion
instance : inhabited (alexandroff X) := ⟨∞⟩
```

justadzr committed:
Update src/topology/alexandroff.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

justadzr committed:
removed an unused argument

urkud committed:
chore(data/option,data/set): a few lemmas, golf

* add `option.subsingleton_iff_is_empty` and an instance
  `option.subsingleton`;
* add `set.is_compl_range_some_none`, `set.compl_range_some`,
  `set.range_some_inter_none`, `set.range_some_union_none`;
* split the proof of `set.pairwise_on_eq_iff_exists_eq` into
  `set.nonempty.pairwise_on_eq_iff_exists_eq` and
  `set.pairwise_on_eq_iff_exists_eq`.

urkud committed:
Merge branch 'master' into alexandroff-extension

urkud committed:
Merge branch 'option-range-some' into alexandroff-extension

urkud committed:
Merge branch 'alexandroff-extension' of git://github.com/leanprover-community/mathlib into alexandroff-extension

urkud committed:
Snapshot

kmill commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,370 @@
+
+@[simp] lemma coe_ne_infty (x : X) : (x : alexandroff X) ≠ ∞  .
+@[simp] lemma infty_ne_coe (x : X) : ∞ ≠ (x : alexandroff X) .
+@[simp] lemma some_eq_coe (x : X) : some x = (x : alexandroff X) := rfl
```
This seems to be an overly far-reaching simp lemma.  For example, it will apply here:
```lean
example (x : ℕ) (z : option ℕ) : some x = z :=
begin
  simp,
  -- now it's (x : alexandroff ℕ) = z
end
```
The next lemma (`none_eq_infty`) has the same issue.

kmill commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,370 @@
+end
+
+/-- If `X` is not a compact space, then `alexandroff X` is a connected space. -/
+lemma connected_space_alexandroff [preconnected_space X] (h : ¬ is_compact (univ : set X)) :
```
Could you mention "This statement cannot be an `instance` because it needs a non-class argument `¬is_compact (univ : set X)`." here, too?

urkud commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,370 @@
+@[simp] lemma coe_ne_infty (x : X) : (x : alexandroff X) ≠ ∞  .
+@[simp] lemma infty_ne_coe (x : X) : ∞ ≠ (x : alexandroff X) .
+@[simp] lemma some_eq_coe (x : X) : some x = (x : alexandroff X) := rfl
+@[simp] lemma none_eq_infty : none = (∞ : alexandroff X) := rfl
```
```suggestion
```

urkud committed:
Update src/topology/alexandroff.lean

kmill commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,370 @@
+/-!
+# The Alexandroff Compactification
+
+We construct the Alexandroff compactification of an arbitrary topological space `X` and prove
```
```suggestion
We construct the Alexandroff compactification (the one-point compactification) of an arbitrary topological space `X` and prove
```
I know the other name is already in the tags, but I think this would make it clearer what this module is for.

justadzr committed:
edited some docstrings

justadzr committed:
Merge branch 'master' into alexandroff-extension

justadzr committed:
Merge branch 'master' into alexandroff-extension

hrmacbeth commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,369 @@
+
+lemma tendsto_nhds_infty {α : Type*} {f : alexandroff X → α} {l : filter α} :
+  tendsto f (𝓝 ∞) l ↔
+    ∀ s ∈ l, f ∞ ∈ s ∧ ∃ t : set X, is_closed t ∧ is_compact t ∧ maps_to (f ∘ coe) tᶜ s :=
```
It would be nice to have versions `tendsto_nhds_infty'` and `continuous_at_infty'` for the more abstract versions using filters, which might even be more useful.  If I understand correctly these should be
```
tendsto f (𝓝 ∞) l ↔ tendsto f (pure ∞) l ∧ tendsto (f ∘ coe) (coclosed_compact X) l
continuous_at f ∞ ↔ tendsto (f ∘ coe) (coclosed_compact X) (𝓝 (f ∞))
```
is that right?

urkud replied:
Done. BTW, could you please write ```` ```lean ... ``` ```` instead of just ```` ``` ... ``` ```` to get syntax highlighting?

justadzr committed:
Merge branch 'master' into alexandroff-extension

justadzr committed:
Merge branch 'master' into alexandroff-extension

urkud committed:
Merge branch 'master' into alexandroff-extension

urkud committed:
Fix

urkud committed:
Fix, restore a lemma

urkud commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,397 @@
+lemma coe_injective : function.injective (coe : X → alexandroff X) :=
+option.some_injective X
+
+@[simp, norm_cast] lemma coe_eq_coe {x y : X} : (x : alexandroff X) = y ↔ x = y :=
```
```suggestion
@[norm_cast] lemma coe_eq_coe {x y : X} : (x : alexandroff X) = y ↔ x = y :=
```

urkud committed:
Update src/topology/alexandroff.lean

urkud committed:
Fix

urkud committed:
Fix

sgouezel commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,397 @@
+We construct the Alexandroff compactification (the one-point compactification) of an arbitrary
+topological space `X` and prove some properties inherited from `X`.
+
+## Main defintions
```
```suggestion
## Main definitions
```

sgouezel commented on src/topology/alexandroff.lean:
```diff
@@ -0,0 +1,397 @@
+## Main defintions
+
+* `alexandroff`: the Alexandroff compactification, we use coercion for the canonical embedding
+  `X → alexandroff X`;
```
Can you add here what happens when `X` is already compact? (I.e., does it add a point, or not?)

justadzr replied:
Done!

sgouezel reviewed (commented):
LGTM, thanks!
And yes, additional API can be added in a later PR.
bors d+

justadzr committed:
Update src/topology/alexandroff.lean

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

justadzr committed:
edited the docstring
